### PR TITLE
fix(parser): reserved binding name in strict mode

### LINKED ISSUE #5120: Strict mode warning not working in entry module
### System Info

Not related

### Details

In webpack, entry module is being translated into an `EntryDependency` and it will be using strict parsing as it's categorized as `Esm`. Rspack does not respect to this option. 

### Reproduce link

_No response_

## Expected
1. Create a demo
2. Add `let let = let` to entry module
3. Run and there should be a parsing error.

## Actual
Rspack does not have a parsing error.

### PR DESCRIPTION
**Pull request by LingyuCoder**
## Summary

Fix #5120 

## Test Plan

- Added `packages/rspack/tests/cases/parsing/issue-5120-xxxx`

## Require Documentation?

- [x] No
- [ ] Yes, the corresponding rspack-website PR is \_\_

### CODE REVIEW

File: crates/rspack_plugin_javascript/src/visitors/dependency/common_js_import_dependency_scanner.rs [hunk 1]
```diff
@@ -232,6 +239,62 @@ impl<'a> CommonJsImportDependencyScanner<'a> {
   pub fn walk_expression(&mut self, expr: &Expr) {
     expr.visit_children_with(self);
   }
+
+  pub fn check_ident(&mut self, ident: &Ident) {
+    if is_reserved_word_in_strict(ident.sym.as_str()) {
+      if self.is_strict {
+        self.errors.push(
+          diagnostic!(
```
Comment (h-a-n-a) [inline crates/rspack_plugin_javascript/src/visitors/dependency/common_js_import_dependency_scanner.rs:247]: Could you please change the diagnostic into a traceable error? This looks more beautiful to user end ;-). You can take this as a reference. https://github.com/web-infra-dev/rspack/blob/bfdc4279f1ae1f1f6332e74db963352d27c012a2/crates/rspack_plugin_javascript/src/utils/mod.rs#L63

File: crates/rspack_plugin_javascript/src/visitors/dependency/common_js_import_dependency_scanner.rs [hunk 2]
```diff
@@ -232,6 +239,62 @@ impl<'a> CommonJsImportDependencyScanner<'a> {
...
+            "SyntaxError: The keyword '{}' is reserved in strict mode",
+            ident.sym
+          )
+          .boxed(),
+        );
+      } else {
+        self.errors.push(
+          diagnostic!(
```
Comment (h-a-n-a) [inline crates/rspack_plugin_javascript/src/visitors/dependency/common_js_import_dependency_scanner.rs:255]: This too ;-)

File: crates/rspack_plugin_javascript/src/visitors/dependency/common_js_import_dependency_scanner.rs (hunk 1 above)
Comment (h-a-n-a) [reply crates/rspack_plugin_javascript/src/visitors/dependency/common_js_import_dependency_scanner.rs:247]: Also I'm planning to move other diagnostic to traceable errors as much as possible.

Comment (LingyuCoder) [reply crates/rspack_plugin_javascript/src/visitors/dependency/common_js_import_dependency_scanner.rs]: Resolved, but only these new errors were modified to traceable, I will modify all the left errors of scanners in another pull request.

File: crates/rspack_plugin_javascript/src/visitors/dependency/common_js_import_dependency_scanner.rs [hunk 3]
```diff
@@ -232,6 +242,60 @@ impl<'a> CommonJsImportDependencyScanner<'a> {
   pub fn walk_expression(&mut self, expr: &Expr) {
     expr.visit_children_with(self);
   }
+
+  pub fn check_ident(&mut self, ident: &Ident) {
+    if is_reserved_word_in_strict(ident.sym.as_str()) {
+      if self.is_strict {
+        self.errors.push(Box::new(create_traceable_error(
+          "SyntaxError".into(),
```
Comment (h-a-n-a) [inline crates/rspack_plugin_javascript/src/visitors/dependency/common_js_import_dependency_scanner.rs:250]: I think you may use 'Javascript parsing error' here to align with other error messages.

**Comment by h-a-n-a** [comment]
Looks almost good to me. But sorry to bother you again. Would you also add a diagnostic test here to avoid future error span regression since it's a graphical diagnostic? `packages/rspack/tests/diagnostics/module-parse-failed`. You may copy other test cases.

**Comment by LingyuCoder** [comment]
> Looks almost good to me. But sorry to bother you again. Would you also add a diagnostic test here to avoid future error span regression since it's a graphical diagnostic? `packages/rspack/tests/diagnostics/module-parse-failed`. You may copy other test cases. Then I will say this PR is good to go :-)

Done

**Review by h-a-n-a — APPROVED** [review]

File: crates/rspack_plugin_javascript/src/visitors/dependency/common_js_import_dependency_scanner.rs [hunk 4]
```diff
@@ -401,13 +465,34 @@ impl<'a> Visit for CommonJsImportDependencyScanner<'a> {
...
+      Program::Script(s) => is_strict(&s.body),
+    };
+    program.visit_children_with(self);
+  }
+
   fn visit_block_stmt(&mut self, n: &BlockStmt) {
     let old_in_strict = self.is_strict;
 
-    self.is_strict = is_strict(n);
+    self.is_strict = is_strict(&n.stmts);
```
Comment (bvanjoi) [inline crates/rspack_plugin_javascript/src/visitors/dependency/common_js_import_dependency_scanner.rs:479]: `blockStmts` should not affect the "use strict" within it, for example this case should work:

```rs
if (true) {
  "use strict"
  let = 1
}
```

**Pull request merged**